### ISSUE TITLE
dart2js incorrectly identifies class name as Modifier

### ISSUE THREAD

**Issue by ajrcarey** [dart-lang/sdk#29072]
dart2js will not generate javascript for the following code fragment:

```
@interface
abstract class State {

  bool containsKey(final String key);

  State remove(final String key);

  State put(final String key, final Object value);

  Optional<T> get<T extends Object>(final String key);

}
```

Instead, it returns the error:

`[Error from Dart2JS on duet|hello-world/index.dart]:
<path>/state.dart:12:3:
Can't have modifier 'Optional' here.
Try removing 'Optional'.
  Optional<T> get<T extends Object>(final String key);
  ^^^^^^^^`

I have no idea what a "Modifier" is, but apparently it thinks Optional is one :)

The class name is not the problem; renaming Optional to something else still results in the dart2js error, just with a different modifier name.

Could it be something to do with the use of function-level generics?

The code passes strong-mode analysis.

**Comment by sigmundch** [comment]
Thanks for filing the bug! and Thank you for reducing this to this small repro, it is super helpful and saves us a lot of time when the repros are this small!

The issue seems to be a combination of the generic argument and the fact that the name of the function is `get` (which is a built-in keyword):

For example: `a.dart` fails, but `b.dart` and `c.dart` are OK:
**a.dart**:
```dart
class State {
  Object get<T>(v) => null;
}
main() => new State();
```

**b.dart**:
```dart
class State {
  Object getState<T>(v) => null;
}
main() => new State();
```

**c.dart**:
```dart
class State {
  Object get(v) => null;
}
main() => new State();
```

**Comment by sigmundch** [comment]
Seems like the issue is in the FE parser (see https://github.com/dart-lang/sdk/blob/c78b4d66bb3e0440fdd3020b45f3d00d78082525/pkg/front_end/lib/src/fasta/parser/parser.dart#L1804)

**Comment by lrhn** [comment]
@peter-ahe-google 
I notified Peter about a similar issue today, he might be working on it already.

**Comment by ajrcarey** [comment]
Ah! I see what you mean. I'm perfectly happy to change my method name. Interesting that dart analyzer doesn't flag it as a problem - does that mean it would work in Dartium? It would be good for all the tools to be in agreement.

**Comment by peter-ahe-google** [comment]
@ajrcarey you're absolutely right, all tools should be in agreement. They haven't been historically, but we're beginning to share a lot of code moving towards the goal of having a single front end shared by all of our tools.

**Comment by jensjoha** [comment]
I can no longer reproduce the error. Marking P3 but leaving it open so we can make sure we have a regression test before closing.

**Comment by ajrcarey** [comment]
Confirmed fixed under 2.0.0-dev.19. Thank you!
